tools/file_packager.py

```diff
@@ -289,6 +289,9 @@ def main():
   }
   Module.expectedDataFileDownloads++;
   (function() {
+   // When running as a pthread, FS operations are proxied to the main thread, so we don't need to
+   // fetch the .data bundle on the worker
+   if (Module['ENVIRONMENT_IS_PTHREAD']) { return; }
```

Comment:
Looking at the code it looks like we actually run the code here by adding it to `preRun` or by calling runWithFS directly in the case that `Module['calledRun']` is set.

Presumably the issue you are seeing must be because Module['calledRun'] is set in the pthreads?

```
    if (Module['calledRun']) {    
      runWithFS();    
    } else {    
      if (!Module['preRun']) Module['preRun'] = [];                                               
      Module["preRun"].push(runWithFS); // FS is not initialized yet, wait for it                 
    }  
```

@kripken I think you have a better understanding of this code than I do.

Whatever we do we should probably try to write a test for this..

Reply:
The overall structure I see is
```
(function() {
  var loadPackage = function() {
    fetchRemotePackage = function(...) {}; // responsible for downloading .data 
    fetchRemotePackage(...); // always called
    runWithFS = function() {...}; // maps downloaded .data to file-system paths
    if (Module['calledRun']) { ... } else { ... } // Call runWithFS() 
  } 
  loadPackage(file_mappings); 
}(); <--- Currently always called on pthread worker
```

The issue is that since the above closure is always called on the worker, we always call `fetchRemotePackage` which will make a request to download the .data file. On worker, this .data file is always assumed to be at the server root, and `getPreloadedPackage` is ignored (as mentioned in the linked issue) which results in a failure to download .data file.

The initial solution I had was to build a mechanism to provide a correctly overloaded `getPreloadedPackage` to the worker and still allow runWithFS() to execute. However, since file-system calls will be proxied from pthread anyways, we can skip the entire `loadPackage` function (per my understanding). Hence, I've modified loadPacakage to early return in case we're running on a pthread worker.

Re: unit test, I was looking to add something, but didn't find any closely matching test. Would be happy to get pointers on how we can test this if not covered by existing tests.

Reply:
I think the convention is to not include { } around the single-line block